prusti-interface/src/environment/borrowck/regions.rs

```diff
@@ -4,33 +4,88 @@
-    lazy_static! {
-        static ref FN_SIG: Regex =
-            Regex::new(r"^fn [a-zA-Z\d_]+\((?P<args>.*)\) -> (?P<result>.*)\{$").unwrap();
+use regex::Regex;
+
+use environment::borrowck::facts;
+use rustc::mir;
+use rustc_data_structures::indexed_vec::Idx;
+
+lazy_static! {
+    static ref FN_SIG: Regex =
+        Regex::new(r"^fn [a-zA-Z\d_]+\((?P<args>.*)\) -> (?P<result>.*)\{$").unwrap();
+}
+lazy_static! {
+    static ref ARG: Regex =
+        Regex::new(r"^_(?P<local>\d+): &'(?P<rvid>\d+)rv (mut)? [a-zA-Z\d_]+\s*$").unwrap();
+}
+lazy_static! {
+    static ref LOCAL: Regex =
+        Regex::new(r"^\s+let( mut)? _(?P<local>\d+): &'(?P<rvid>\d+)rv ").unwrap();
+}
+lazy_static! {
+    static ref LOCAL_TUPLE: Regex =
+        Regex::new(r"^\s+let( mut)? _(?P<local>\d+): \((?P<items>.*)\);").unwrap();
+}
+lazy_static! {
+    static ref REF: Regex = Regex::new(r"&'(?P<rvid>\d+)rv ").unwrap();
+}
+
+#[derive(Debug)]
+pub struct PlaceRegions(HashMap<(mir::Local, Vec<usize>), facts::Region>);
+
+impl PlaceRegions {
+    fn new() -> Self {
+        PlaceRegions(HashMap::new())
+    }
+
+    fn add_local(&mut self, local: mir::Local, rvid: facts::Region) {
+        self.add(local, vec![], rvid);
+    }
+
+    fn add(&mut self, local: mir::Local, projections: Vec<usize>, rvid: facts::Region) {
+        self.0.insert((local, projections), rvid);
+    }
+
+    pub fn for_local(&self, local: mir::Local) -> Option<facts::Region> {
+        self.for_place(&mir::Place::Local(local))
     }
-    lazy_static! {
-        static ref ARG: Regex =
-            Regex::new(r"^_(?P<local>\d+): &'(?P<rvid>\d+)rv (mut)? [a-zA-Z\d_]+\s*$").unwrap();
+
+    /// Determines the region of a MIR place. Right now, the only supported places are locals and tuples. Tuples cannot be nested inside other tuples.
+    pub fn for_place(&self, place: &mir::Place) -> Option<facts::Region> {
+        let (local, fields) = Self::translate_place(place);
+        self.0.get(&(local, fields)).cloned()
     }
-    lazy_static! {
-        static ref LOCAL: Regex =
-            Regex::new(r"^\s+let( mut)? _(?P<local>\d+): &'(?P<rvid>\d+)rv ").unwrap();
+
+    /// Translates a place like _3.0.3.1 into a local (here _3) and a list of field projections like (here [0, 3, 1]).
+    fn translate_place(place: &mir::Place) -> (mir::Local, Vec<usize>) {
+        match place {
+            mir::Place::Local(local) => (*local, vec![]),
+            mir::Place::Projection(box projection) => {
+                let (local, mut fields) = Self::translate_place(&projection.base);
+                match projection.elem {
+                    mir::ProjectionElem::Field(f, _) =>
+                        fields.push(f.index()),
+                    _ => unreachable!()
+                };
+                (local, fields)
+            },
+            _ => unreachable!("unsupported place: {:?}", place)
+        }
     }
+}
+
+pub fn load_place_regions(path: &Path) -> io::Result<PlaceRegions> {
```

Comment:
I changed this to load the regions not only for local variables, but also for places inside tuples. So if there is a local variable like this:

    let _5: (&'6rv mut i32, &'7rv mut i32),

then it remembers that the place `_5.0` has region 6 and the place `_5.1` has region 7. Internally, places are represented as a local variable plus a vector of field projections (because I can't construct the `mir::Place` without type information), but the `PlaceRegions` struct (that stores the region information) provides a way to query the region for a given `mir::Local` or `mir::Place`.